### PR TITLE
Unifies DPC record `program_id`, introduces `ProgramIDTreeParameters`

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

This PR unifies birth and death programs in DPC records into a `program_id`.

- Introduces `ProgramIDTreeParameters`
- Introduces a `ProgramCircuitTree` struct as an internal helper for all programs.
- Introduces `Program` and `ProgramCircuit` traits
- Introduces a `ProgramPrivateVariables` trait to be implemented for each circuit.

## Considerations

This  PR introduces `Send + Sync` for many `SNARK` associated types.

An open question with the unified program ID model is whether we need to enforce the `circuit_index` as part of the local data commitment. At the moment, the local data commitment is unchanged.

## Related PRs

This PR replaces #324.

### CODE REVIEW

**Comment by howardwu** [comment]
This PR will require resampling all parameters. I will perform the sampling after approvals.

**Comment by howardwu** [comment]
As a sanity check, this PR and PR #324 both have the same number of constraints for the outer circuit. So the steps performed to derive the `program_id` should be equivalent.

File: dpc/src/record/record.rs [hunk 1]
```diff
@@ -59,73 +55,59 @@ pub struct Record<C: Parameters> {
...
         rng: &mut R,
     ) -> Result<Self, RecordError> {
-        let record_time = start_timer!(|| "Generate record");
-        // Sample new commitment randomness.
-        let commitment_randomness = <C::RecordCommitmentScheme as CommitmentScheme>::Randomness::rand(rng);
+        let timer = start_timer!(|| "Generate record");
+        let program_id = program.program_id().to_bytes_le()?;
 
-        // Total = 32 + 1 + 8 + 128 + 48 + 48 + 32 = 297 bytes
+        // Total = 48 + 32 + 1 + 8 + 128 + 48 + 32 = 297 bytes
```
Comment (ljedrz) [inline dpc/src/record/record.rs:89]:
```suggestion
        // Total = 48 + 32 + 1 + 8 + 128 + 32 = 297 bytes
```

**Review by ljedrz — APPROVED** [review]
LGTM with one comment tweak

**Review by raychu86 — APPROVED** [review]

**Comment by howardwu** [comment]
Resampling all parameters now

**Pull request merged**